include/unitTestMain.h

```diff
@@ -21,13 +21,42 @@
 #include <cppunit/TestResultCollector.h>
 #include <cppunit/TestRunner.h>
 
+/// implements test program's main() function while enabling customization
+class TestProgram
```

Comment:
Classes representing basic concepts like "a program" or "an application" are very common indeed. New Squid developers reading this code should recognize familiar patterns.

> FME; What has changed since my almost identical proposal was vetoed (circa 2008) as `Bad Practice`?

I cannot answer that question without seeing that (proposal and its) veto. FWIW, I glanced through Subject lines of squid-dev postings from 2007 to 2009 and could not find anything relevant mentioning "application" or "program"; I also looked through same Subject lines mentioning "RFC". Needless to say, this search is not accurate or exhaustive, and I probably just missed your proposal. Can you share a pointer?

Reply:
Sorry I got the year wrong. Latest proposal was 19 June 2017 "[PATCH] helper process library" - mailing list mirrors appear to have dropped the patch from my latest revision.
If you are okay with this MFC design now I can revamp the old patch and open a PR proposing its use for helpers alongside this one for test binaries.

Reply:
> Latest proposal was 19 June 2017 "[PATCH] helper process library"

AFAICT, my [review](http://lists.squid-cache.org/pipermail/squid-dev/2017-June/009003.html) of that [proposal](http://lists.squid-cache.org/pipermail/squid-dev/2017-June/009002.html) matches this PR very well -- this PR fixes the _same_ main() problems (in the official unit testing code) that 2017 proposal would have introduced into Squid helper code and that my 2017 review has identified.

For example, the TestProgram class implemented in this PR matches the Helper::Instance or Helper::Application class fix suggested in my 2017 review, and the following main() function sketched in my 2017 review perfectly matches the main() functions added in this PR:

```C++
// 2017 review sketch:
int
main(int argc, char *argv[])
{
    return MyInstance().run(argc, argv);
}
```

BTW, instead of proposing a Program or Application class, that 2017 [patch](//lists.squid-cache.org/pipermail/squid-dev/attachments/20170619/13fb64c1/attachment-0001.bin) proposed a C++ _namespace_:

```C++
// 2017 patch (with comments stripped):
namespace HelperProcess {
extern std::string Name;
bool Start();
bool Options(int argc, char *argv[]);
void DoOneRequest(std::string &buffer);
} // namespace HelperProcess
```


> What has changed since my almost identical proposal was vetoed as `Bad Practice`?

I can now answer that question. Nothing has changed at the high-level AFAICT: 2017 proposal had serious architectural flaws that were identified in my 2017 review, with specific suggestions on how to address them. This PR does not have those flaws.

> If you are okay with this MFC design now

The Microsoft Foundation Class library ([MFC](https://learn.microsoft.com/en-us/cpp/mfc/mfc-application-architecture-classes?view=msvc-170)) design is a red herring: It was not proposed or reviewed in 2017. It is not under review here and now. The TestProgram class introduced in this PR has similarities with MFC [CWinApp](https://learn.microsoft.com/en-us/cpp/mfc/reference/cwinapp-class?view=msvc-170) class and with numerous "application" or "program" classes in numerous other libraries and programs. There was and still is nothing wrong with basic Program or Application class design per se, but it was not a part of that 2017 proposal.


> I can revamp the old patch and open a PR proposing its use for helpers

If you resume working on that helper refactoring, please address my 2017 concerns before posting a PR. Most of them are critical to getting the overall design right. Also, to avoid redoing a lot of changes, please do not convert more than one helper until there is an agreement that the PR is on the right track (but be prepared to convert them all at the end of the process).